compiler/rustc_span/src/lev_distance.rs

```diff
@@ -1,49 +1,104 @@
-//! Levenshtein distances.
+//! Damerau-Levenshtein distances.
 //!
-//! The [Levenshtein distance] is a metric for measuring the difference between two strings.
+//! The [Damerau-Levenshtein distance] is a metric for measuring the difference between two strings.
+//! This implementation is a restricted version of the algorithm, as it does not permit modifying
+//! characters that have already been transposed.
 //!
-//! [Levenshtein distance]: https://en.wikipedia.org/wiki/Levenshtein_distance
+//! [Damerau-Levenshtein distance]: https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance
 
 use crate::symbol::Symbol;
-use std::cmp;
+use std::{cmp, mem};
 
 #[cfg(test)]
 mod tests;
 
-/// Finds the Levenshtein distance between two strings.
+/// Finds the restricted Damerau-Levenshtein distance between two strings. Characters that have
+/// already been transposed may not be modified.
 ///
 /// Returns None if the distance exceeds the limit.
 pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
-    let n = a.chars().count();
-    let m = b.chars().count();
-    let min_dist = if n < m { m - n } else { n - m };
+    let mut a = a.chars().collect::<Vec<_>>();
+    let mut b = b.chars().collect::<Vec<_>>();
+
+    // Ensure that `b` is the shorter string, minimizing memory use.
+    if a.len() < b.len() {
+        mem::swap(&mut a, &mut b);
+    }
 
+    let min_dist = a.len() - b.len();
+    // If we know the limit will be exceeded, we can return early.
     if min_dist > limit {
         return None;
     }
-    if n == 0 || m == 0 {
-        return (min_dist <= limit).then_some(min_dist);
-    }
 
-    let mut dcol: Vec<_> = (0..=m).collect();
+    // Strip common prefix.
+    loop {
+        match a.split_first().zip(b.split_first()) {
+            Some(((a_char, a_rest), (b_char, b_rest))) if a_char == b_char => {
+                a = a_rest.to_vec();
+                b = b_rest.to_vec();
+            }
+            _ => break,
+        }
+    }
+    // Strip common suffix.
+    loop {
+        match a.split_last().zip(b.split_last()) {
+            Some(((a_char, a_rest), (b_char, b_rest))) if a_char == b_char => {
+                a = a_rest.to_vec();
+                b = b_rest.to_vec();
+            }
+            _ => break,
+        }
+    }
 
-    for (i, sc) in a.chars().enumerate() {
-        let mut current = i;
-        dcol[0] = current + 1;
+    // If either string is empty, the distance is the length of the other.
+    // We know that `b` is the shorter string, so we don't need to check `a`.
+    if b.len() == 0 {
+        return Some(min_dist);
+    }
 
-        for (j, tc) in b.chars().enumerate() {
-            let next = dcol[j + 1];
-            if sc == tc {
-                dcol[j + 1] = current;
-            } else {
-                dcol[j + 1] = cmp::min(current, next);
-                dcol[j + 1] = cmp::min(dcol[j + 1], dcol[j]) + 1;
+    // Variable length arrays are not yet implemented.
```

Comment:
In C there is advice against using VLAs due to optimization and security reasons, with the kernel [dropping them completely](https://www.phoronix.com/news/Linux-Kills-The-VLA).

If you care about performance, you could use smallvecs here. I think if you put 16 characters as the limit that ought to be enough to not require allocations for 95% of literals out there.

In general this code should be rather cold though as it is only executed for diagnostics which the rustc compiler doesn't emit *that* many.

Reply:
I was not aware of that advice. I'll remove that comment.

This code should definitely be cold, which is why I didn't focus too much on eliminating the initial allocation of `Vec<char>`. But I was bored, so I figured let's do something useful.

`SmallVec` is something I considered. With the suggested changes elsewhere, I think it'll have minimal impact, though.